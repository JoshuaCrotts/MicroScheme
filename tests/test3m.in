
// Sieve...  just for something else to do...

char flags[100];

void strcpy(char[] dst, char[] src) {
    int i;
    for (i=0; i<#src && i<#dst; i++) {
        dst[i] = src[i];
        if (dst[i] == '\0')
            break;
    }
}

int isprime(int x) {
    return flags[x] == 0;
}

void checkprime(int x) {
	char buf[10];
    printd(x);
    if (isprime(x)) 
    	strcpy(buf, " is ");
    else 
    	strcpy(buf, " is not ");
    
    prints(buf);
    prints("prime\n");
}

void main() {
    int i;
    int j;
    for (i=0; i<#flags; i++)
        flags[i] = 0;
        
    for (i=2; i<#flags; i++) {
        if (!flags[i]) {
            for (j=2*i; j<#flags; j=j+i)
                flags[j] = 1;
        }
    }
    
    checkprime(8);
    checkprime(29);
    checkprime(51);
    checkprime(53);
}
