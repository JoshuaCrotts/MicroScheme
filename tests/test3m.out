8 is not prime
29 is prime
51 is not prime
53 is prime
